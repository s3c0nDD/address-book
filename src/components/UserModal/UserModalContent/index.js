import UserModalContent from './UserModalContent';

export default UserModalContent;