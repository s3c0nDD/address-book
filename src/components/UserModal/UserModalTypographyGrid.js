import React from 'react';

import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';

const UserModalTypography = ({ text }) => (
  <Typography
    variant="body2"
    color="textSecondary"
    component="p"
  >
    {text}
  </Typography>
);

const UserModalTypographyGrid = ({ IconComponent, CustomContent, text, ...props }) => {
  return (
    <Grid container {...props}>
      <Grid item xs={2}>
        { IconComponent }
      </Grid>
      <Grid item xs={10}>
        {text
          ? <UserModalTypography text={text} />
          : CustomContent
        }
      </Grid>
    </Grid>
  );
};

export default UserModalTypographyGrid;

export {
  UserModalTypography
}