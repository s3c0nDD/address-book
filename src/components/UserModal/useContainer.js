import { useDispatch, useSelector } from 'react-redux';

import { openModal, closeModal } from '../../store/module';

const useContainer = () => {
  const dispatch = useDispatch();

  const user = useSelector(state => state.app.modalUser);
  const isOpen = Boolean(user);

  const doOpenModal = (user) => dispatch(openModal(user));
  const doCloseModal = () => dispatch(closeModal());

  return [{ user, isOpen }, { doOpenModal, doCloseModal }]
};

export default useContainer;