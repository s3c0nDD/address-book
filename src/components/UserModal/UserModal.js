import React, { memo, useMemo } from 'react';
import { makeStyles } from '@material-ui/core/styles';

import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import EmailIcon from '@material-ui/icons/Email';
import HomeIcon from '@material-ui/icons/Home';
import PersonIcon from '@material-ui/icons/Person';
import PhoneAndroidIcon from '@material-ui/icons/PhoneAndroid';
import PhoneIcon from '@material-ui/icons/Phone';

import UserModalTypographyGrid, { UserModalTypography } from './UserModalTypographyGrid';

const useStyles = makeStyles(theme => ({
  card: {
    borderRadius: 0
  },
  media: {
    height: 200,
    [theme.breakpoints.up('sm')]: {
      height: 260,
    },
    [theme.breakpoints.up('md')]: {
      height: 300,
    },
    [theme.breakpoints.up('lg')]: {
      height: 340,
    }
  },
  content: {
    margin: theme.spacing(1),
    [theme.breakpoints.up('sm')]: {
      margin: theme.spacing(2),
    },
  }
}));

const shouldNotRerender = (prevProps, nextProps) => {
  return prevProps.user === nextProps.user;
};

const UserModal = memo(({ user, onCloseModal }) => {
  const classes = useStyles();

  const isOpen = useMemo(() => Boolean(user), [user]);
  const fullNameWithTitle = useMemo(() => {
    return `${user?.name?.title} ${user?.name?.first} ${user?.name?.last}`;
  }, [user]);

  if (!user) { return null; }

  return (
    <Dialog
      fullWidth={true}
      maxWidth={'xs'}
      onClose={onCloseModal}
      aria-labelledby="user-dialog-title"
      open={isOpen}
    >
      <DialogTitle id="user-dialog-title">
        Details
      </DialogTitle>

      <Card className={classes.card}>
        <CardMedia
          className={classes.media}
          image={user.picture.large}
          title={`${fullNameWithTitle} face photo`}
        />
        <CardContent className={classes.content}>
          <Typography
            gutterBottom
            variant="h4"
            component="h2"
          >
            {fullNameWithTitle}
          </Typography>

          <UserModalTypographyGrid
            IconComponent={<PersonIcon color="primary" />}
            text={user.login.username}
          />

          <UserModalTypographyGrid
            IconComponent={<EmailIcon color="primary" />}
            text={user.email}
          />

          <UserModalTypographyGrid
            IconComponent={<PhoneIcon color="primary" />}
            text={user.phone}
          />

          <UserModalTypographyGrid
            IconComponent={<PhoneAndroidIcon color="primary" />}
            text={user.cell}
          />

          <UserModalTypographyGrid
            alignItems="center"
            IconComponent={<HomeIcon color="primary" />}
            CustomContent={
              <Grid
                container
                direction="column"
              >
                <Grid item xs>
                  <UserModalTypography
                    text={`${user.location.street.name} ${user.location.street.number}`}
                  />
                </Grid>
                <Grid item xs>
                  <UserModalTypography
                    text={`${user.location.postcode} ${user.location.city} `}
                  />
                </Grid>
                <Grid item xs>
                  <UserModalTypography
                    text={`${user.location.state} / ${user.location.country}`}
                  />
                </Grid>
              </Grid>
            }
          />

        </CardContent>
      </Card>
    </Dialog>
  );
}, shouldNotRerender);

export default UserModal;
