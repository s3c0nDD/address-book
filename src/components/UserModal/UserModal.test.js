import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import userMockData from '../mocks/userMock'
import UserModal from './UserModal'

describe('UserModal', () => {
  let shallow;
  let wrapper;
  let closeModalMock;
  let userMock = { ...userMockData };

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    closeModalMock = jest.fn();
  });

  it('shallowly renders and matches snapshot', () => {
    wrapper = shallow(
      <UserModal
        user={userMock}
        onCloseModal={closeModalMock}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render null when no prop user provided', () => {
    wrapper = shallow(
      <UserModal
        user={null}
        onCloseModal={closeModalMock}
      />
    );
    expect(wrapper.html()).toBeNull();
  });
});