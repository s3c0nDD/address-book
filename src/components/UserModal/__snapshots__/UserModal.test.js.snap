// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserModal shallowly renders and matches snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="user-dialog-title"
  fullWidth={true}
  maxWidth="xs"
  onClose={[MockFunction]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="user-dialog-title"
  >
    Details
  </WithStyles(ForwardRef(DialogTitle))>
  <UserModalContent
    user={
      Object {
        "cell": "111-222-333",
        "email": "johndoe123@google.com",
        "id": Object {
          "value": "foobar",
        },
        "location": Object {
          "city": "Warsaw",
          "country": "Poland",
          "postcode": 56789,
          "state": "Masovian",
          "street": Object {
            "name": "Big Street",
            "number": 98,
          },
        },
        "login": Object {
          "username": "johndoe123",
        },
        "name": Object {
          "first": "John",
          "last": "Doe",
          "title": "Mr",
        },
        "phone": "000-111-222",
        "picture": Object {
          "large": "https://wwww.image.com/data/1",
        },
      }
    }
  />
</WithStyles(ForwardRef(Dialog))>
`;
